src/passes/OptimizeInstructions.cpp

```diff
@@ -841,6 +757,54 @@ struct OptimizeInstructions
           }
         }
       }
+      if (auto* constTrue = select->ifTrue->dynCast<Const>()) {
+        if (auto* constFalse = select->ifFalse->dynCast<Const>()) {
+          if (select->type == Type::i32 || select->type == Type::i64) {
+            auto trueValue = constTrue->value.getInteger();
+            auto falseValue = constFalse->value.getInteger();
+            if ((trueValue == 1LL && falseValue == 0LL) ||
+                (trueValue == 0LL && falseValue == 1LL)) {
+              Builder builder(*getModule());
+              // canonicalize "expr ? 0 : 1" to "!expr ? 1 : 0"
+              if (trueValue == 0LL) {
+                select->condition = optimizeBoolean(
+                  builder.makeUnary(EqZInt32, select->condition));
```

Comment:
if the true value is 0, the select's condition is flipped. but the arms are not - that seems like a bug?

also we must check if we can flip the arms - side effects might prevent that. `canReorder` helps with that.

Reply:
> if the true value is 0, the select's condition is flipped. but the arms are not - that seems like a bug?

No, due to arms as constants and not taken into account (we return only `select->condition` part)

Reply:
I see, thanks.

This is confusing, though: the comment says `// canonicalize "expr ? 0 : 1" to "!expr ? 1 : 0"` but that is not what it is doing, as I understand the comment. The 0 and 1 values are not moved around.

I'll write some suggestions.